### ISSUE TITLE
save utf-8 with BOM?

### ISSUE THREAD

**Issue by laramjo** [adobe/brackets#10583]
Hello. I want to save a document in UTF8 with BOM ...
How do I?

Guardo document and tittle I get strange characters

**Comment by peterflynn** [comment]
Placing a BOM in a UTF-8 file is specifically recommended _against_ by the Unicode standard: http://stackoverflow.com/a/2223926/1172352.  So personally I don't think Brackets should do work to give users a way to do this.  (Presumably a command-line tool or Brackets extension could be used to add the BOM in special cases where it's needed for some reason).

> save the document and it shows strange characters

Are you talking about opening a file saved by Brackets in some other app?  What app are you reading the file with?  It sounds like that app may not support UTF-8...

**Comment by laramjo** [comment]
Ok. thank you very much for the info. Might help.

Except my project, I close it and reopen it I get some letters and strange characters, especially where words have accent.

look at an example:

Este exto es la creaciÃ³n de mÃ­ probia autorÃ­a

google translation

**Comment by peterflynn** [comment]
@laramjo Where are you getting that text from?  Are you pasting it in from another app or webpage?  Is it a public page where I could try copy-pasting the same text on my machine to replicate this?

**Comment by HybridSolutions** [comment]
I'm also experiencing this when editing .aspx webpages with special characters or words with accents. They look great in Visual Studio, Notepad2, Notepad++ and after saving those files in each of these editors, the pages show the correct text in any browser. If I open them with Brackets, edit and save, the special characters get all messed up but only in the browser. To correct the files, I have to open again in Notepad2, change encoding to UTF with signature and save them. This is happening since the first version of Brackets and should be corrected.

**Comment by reginas** [comment]
I have also experienced this issue, specifically when editing and saving a .json data file that has diacritic (accent marks, etc) information. The json file is imported through .jsx for scripting photoshop. All characters with accent marks in the .json file (that has been edited and saved with brackets) end up as checkmarks, degree symbols, etc. in Photoshop. When I open/save with a different text editor, such as TextWrangler, and restore the "with BOM" flavor of UTF-8 encoding, the characters are back to normal on the next import into Photoshop. I've attached example images.

<img width="321" alt="data_from_json_utf-8-with-bom" src="https://cloud.githubusercontent.com/assets/101154/9837147/81e4b5fc-59fb-11e5-9a4f-3a8d033f0c72.png">
<img width="300" alt="data_from_json_utf-8-bom-was-stripped-by-brackets" src="https://cloud.githubusercontent.com/assets/101154/9837148/81e6748c-59fb-11e5-89fd-f044171da0f2.png">

**Comment by reginas** [comment]
After reading through the recommendations against automatically including BOM, I can see the problems it may cause, but it would be nice to have the OPTION to include BOM as an application preference, or even a plugin, so that people using any applications that seem to rely on Bit Order Marker in their workflow which includes brackets, can still use brackets. 

I specifically use brackets because it has so many great features that TextWranler, Notepad ++, etc. do not have, so again, a big thanks for all of you working to make and keep brackets so useful to us all.

**Comment by reginas** [comment]
Listing items apps/uses that *MAY* have issues with NO byte order mark, as a clue to investigate, (not definitive, just gleaned from stack overflow comments:

Excel and some other MS products [excluding SQL*Plus and VBScript, which prefer NO BOM]
Photoshop
.json files imported into some apps
.xml files imported into some apps
.asp/.aspx components
Visual C++ compiler
txt, xml, csv in Windows or Unix, initially ASCII then had UTF-8 content added
Visual Studio
SSMS
SoureTree

**Comment by neilolee** [comment]
brackets = not +BOM unicode = 
https://www.freeliving.com.tw/i-About-Us-brackets.aspx


Sublime Text = Optional +BOM or -BOM unicode =
 https://www.freeliving.com.tw/i-About-Us.aspx


Brackets Can increase as Sublime Text features?
![code](https://cloud.githubusercontent.com/assets/9427911/11085491/f84f8596-8881-11e5-8990-6eb93d59561e.png)

**Comment by HybridSolutions** [comment]
Really don't understand why Brackets still is missing this one! I use Notepad++, Notepad2, Visual Studio Code, among others, and they all can deal with files with special characters. Why don't you try to edit an html file with text using ç, ã, õ and see what happens. Fix it, please!

**Comment by nikitogr** [comment]
It's very sad that bracket devs don't want to cooperate and listen to its users' voices. It's very serious, and we're not talking about adding a feature. It's vital because at this moment, brackets for me is just a disappointing IDE  and it's just unusable. Until devs change their minds, I won't recommend brackets at all. I strictly use greek characters and brackets fails to meet the basic things. Imagine if it was payware, this would be hilarious. It's alright to use brackets as a toy and playing around. Till then, I'm gonna use serious programs. Change your minds to change ours. Sure, you got no profit (probably) but it doesn't hurt to listen us.

**Comment by gcacars** [comment]
Same here... but BOM breaks PHP files. I think that Brackets should always put BOM if the file is a .cshtml or .aspx and he should always remove BOM if the file is PHP.

**Comment by Download** [comment]
@peterflynn 
> Placing a BOM in a UTF-8 file is specifically recommended against by the Unicode standard

This is not true. They 'require nor recommend it'. That is not the same as recommending *against* it. The reason they do not recommend it (or did not, the source is old) is that programs that are *not capable of dealing with Unicode* might get confused by it.

However, at the same time, they do require implementations to support reading files with a BOM:

> When a process interprets a byte sequence which purports to be in a Unicode character
encoding scheme, it **shall** interpret that byte sequence according to the byte order and
specifications for the use of the byte order mark established by this standard for that
character encoding scheme.
(<sub><sup>[Unicode Standard 5.0, ch.3, C11](http://www.unicode.org/versions/Unicode5.0.0/ch03.pdf), emphasis mine)

This means that all applications that *do support Unicode* need to accept the BOM.

The problem is that without the BOM, applications that accept text files have no way of knowing what encoding the content is in. Just try this: Create a new Unicode file with Brackets and put some content in it, but limit characters used to ASCII. Save it. Now download Notepad++ (which is a fantastic editor BTW) and open the file. Look in menu `Encoding`. What does it say?

It says ANSI.

Notepad++ has no way of knowing the content encoding so it has to guess. It sees only ASCII so it assumes the lowest common denominator (which basically on Windows is ASCII + foreign language extensions, e.g. [Windows-1252](https://en.wikipedia.org/wiki/Windows-1252)).

The problem arises when we start to edit in characters that are outside of ASCII... Characters such as `'€'`. They will be saved in the wrong encoding. I know Notepad++ does this (and I wish it didn't) but there definitely are many editors that do this. But Notepad++ *does support Unicode*! We just have to make sure it does not have to guess the encoding. We do so by... adding a BOM.

Please, pretty please, prepend the BOM to all files on saving. Or at least give us an option to do so. It is such a huge help when the whole toolchain can *know* a file's content is Unicode instead of having to *guess* it. And as it is a requirement for Unicode support, it is required to be supported by all programs that support Unicode. 

Also see my comments on related issue #3898

**Comment by Download** [comment]
@HybridSolutions 
> I'm also experiencing this when editing .aspx webpages with special characters or words with accents. They look great in Visual Studio, Notepad2, Notepad++ and after saving those files in each of these editors, the pages show the correct text in any browser. If I open them with Brackets, edit and save, the special characters get all messed up but only in the browser.

The reason you are experiencing this is that Bracket strips the BOM when you save the file. Apparently you are not sending the correct `Content-Type` header: `text/html; charset=UTF-8`. Probably you are opening the file directly from the file system (`file://...` url)? If the file has a BOM, the browser can infer the encoding from there. If not, it has to guess and it does so by examining the first `n` (512?) bytes of the content. You can probably work around it by specifying the encoding in a meta tag: `<meta charset="utf-8" />`, but it would be waaay better if Brackets would write out the BOM so the browser doesn't have to guess in the first place. Same goes for all the editors and Visual Studio. They all have to guess the encoding because Brackets does not write out the BOM.

**Comment by gcacars** [comment]
No. Visual Studio (or maybe IIS) don't works right even with `Content-Type` nor `<meta charset="utf-8" />`. He needs the BOM.
![vsutf](https://cloud.githubusercontent.com/assets/1190940/12204760/56fe69ea-b61d-11e5-9cbd-5a0c7885d846.PNG)

**Comment by Download** [comment]
@gcacars Mmm. I'm pretty sure that if the content is in fact Unicode/UTF-8 *and* you set the headers correctly (as shown in the screenshot), it should work... So check your content... Is it really UTF-8?

But... Point is, this would all be much simpler if we did set the BOM. So yeah, hoping the Brackets developers pick up on this.

**Comment by HybridSolutions** [comment]
@Download, here's some screenshots. As you can see I have correct headers set, and I'm not opening a local file. This issue should be corrected or brackets is useless.

![image](https://cloud.githubusercontent.com/assets/8672899/12446650/23ab3556-bf63-11e5-979c-fa6a6a97b34a.png)

![image](https://cloud.githubusercontent.com/assets/8672899/12446616/f137e31c-bf62-11e5-90f8-4ca09bb38f16.png)

As you can see, that button is not showing the last word in a correct way. If I open this same file with any of the editors I mentioned previously, and save again, the problem is fixed.

**Comment by Download** [comment]
@HybridSolutions I agree this issue should be resolved by supporting the BOM (imho, ideally, it would always write out a BOM unless explicitly told not to), but I *still* think your problem is unrelated.

Browsers do not need a BOM to correctly display HTML as Unicode when given the correct content type headers (as you are indeed doing) so I still think that either your content is not in fact Unicode, or you have some server settings wrong.

One exception might be if your HTML does not start with `<DOCTYPE html>`; you might end up in some compatibility mode in which the browser assumes latin... But honestly I'd be surprised by it. Fact remains that you do not need a BOM to get a working webpage if you do everything else right. 

Seeing that you are using IIS and Microsoft hasn't always been too Unicode friendly, if you are sure your content is in fact valid UTF-8, I'd check whether IIS is making assumptions about it when reading the source file; it could be thinking it's reading latin, then converting to UTF-8 before sending it to the browser. I'm not into IIS but I'd check if it has settings for the default content type to assume for files that do not have a BOM and what they are set to.

The last scenario, is a great summary of the issues that come with not writing out the BOM; all applications that are reading such files have to *guess* at the encoding because there is no way to infer it from te environment. IIS is reading from the file system so it has no way to know for sure what content type the source file is in. Unicode has an elegant solution built into it: the BOM. Let's use it!

**Comment by Download** [comment]
Looking closer at your screenshot I am now convinced IIS is reading your source file as latin (interpreting multi-byte characters as multiple characters, giving the signature capital A with tilde followed by some other character). Then, when it sends it to the browser, it converts it to UTF-8, but the damage was already done so you get a UTF-8 encoded version of the already broken page. The BOM would have prevented IIS from reading it as latin, but you can almost certainly tell IIS explicitly to assume UTF-8 instead. Not that it makes this issue any less annoying, but I do believe there *is* a fix (workaround) for you that does not involve waiting around for this issue to be resolved.

**Comment by TabrisK** [comment]
A workaround can fix this problem on windows.After created a file in brackets,find it and open with notepad,then click file -> save as.. -> change the encoding as UTF-8 before save the file.

**Comment by HybridSolutions** [comment]
Officially abandoned Brackets! Using Visual Studio Code Free editor! Once again, the issue with brackets, is a non issue with this new editor as it is a non issue with Notepad++ and several others. Wake up Brackets Team!

**Comment by brendonmm** [comment]
@download, the problem related for others in this issue isn't only by file system access, but in all files after have been edited by Brackets. I really love to use as default editor, but, always I open aspx or cshtml files, before encoded/created by Visual Studio, all of the characters loaded on browser began wrong. 

I tried this on latest Chrome/Firefox/Edge browsers, with IIS 10 or IIS 7.5 Express (my older, loaded from Webmatrix). 

If I live preview from Brackets the UTF-8 characters load correctly, but, loading the same page on server-side IIS, I see wrong chars. 

In other tests, JSON files created on Notepad++ as UTF-8 won't be open in Brackets till I reopen that on N++ and convert that to UTF8 w/ BOM.

**Comment by Download** [comment]
@brendonmm Yeah I've found that this issue makes Brackets unusable for me so I've gone back to Notepad++ for now. Still keeping an eye on this issue in the hopes they will fix it but not holding my breath.

**Comment by kasper3** [comment]
Well I rather switched to VSCode myself. BTW @peterflynn, every editor I have used respects the BOM if it is present and provides a way to add it if missing. No editor other than Notepad "forces" the BOM down the throats. I am not sure why that one SO post, that is full of contradictory interpretations, has given importance over the real world demand.

#### WHAT DOES `UTF8 WITH BOM` MEAN TO AN "EDITOR"?

It is a requirement.

#### IS BOM DYING FROM THE FACE OF THE EARTH?

1. All modern major interpreters, compilers and linters support it. Nobody is particualarly dropping the support. New linters for JS superset languages forexample, are very recent and yet support BOM. So nobody is being paranoid like that standard itself..
2. Shebang interpretation is done by shell (ash,bash,csh,dash,tsh etc.) and not the underlying runtime. User can add a preprocessor to strip away BOM. It's NOT a 'there is no-way-out scenario' if you happen to have scripts stored with BOM.
3. JSON parsers in browser chose not to ignore BOM due to standard abidance. The same standarad says that you cannot have code comments in JSON (and how many people like that part). Anyway, BOM in JSON files saved in storage is not vehemently unusual. When transmitted over the wire, the applications strip it off for browser.

So yeah maybe folks are trying to kill BOM, like Apple is trying to kill Adobe, but both are surviving and for an editor, it is a MUST HAVE REQUIREMENT.
